ipn/ipnlocal: panic: Log in goroutine after TestStateMachine has completed

Test flake at 47044f3af7cf96332cddd2c6052df3fe9364d271 ....

@apenwarr, can you look?

```
panic: Log in goroutine after TestStateMachine has completed: 
(ignored) Notify{wg={0 0 0 0 map[]}}



goroutine 156 [running]:
testing.(*common).logDepth(0xc0003e01a0, {0xc0003b5f80, 0x27}, 0x3)
	/opt/hostedtoolcache/go/1.17.1/x64/src/testing/testing.go:779 +0x4c9
testing.(*common).log(...)
	/opt/hostedtoolcache/go/1.17.1/x64/src/testing/testing.go:761
testing.(*common).Logf(0x40c5ab, {0xb22f8f, 0xc000291c00}, {0xc0000466b0, 0xc000291c00, 0xc000069cf8})
	/opt/hostedtoolcache/go/1.17.1/x64/src/testing/testing.go:807 +0x4c
tailscale.com/ipn/ipnlocal.TestStateMachine.func2({{}, {0xb21c02, 0xd}, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc0002364e0, 0x0, ...})
	/home/runner/work/tailscale/tailscale/ipn/ipnlocal/state_test.go:318 +0x1e3
tailscale.com/ipn/ipnlocal.(*LocalBackend).send(0x0, {{}, {0xb21c02, 0xd}, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc0002364e0, ...})
	/home/runner/work/tailscale/tailscale/ipn/ipnlocal/local.go:1207 +0x13c
tailscale.com/ipn/ipnlocal.(*LocalBackend).setWgengineStatus(0xc0004026c0, 0xc00016c2c0, {0x0, 0x0})
	/home/runner/work/tailscale/tailscale/ipn/ipnlocal/local.go:630 +0x375
tailscale.com/wgengine.(*userspaceEngine).RequestStatus(0xc0003fe000)
	/home/runner/work/tailscale/tailscale/wgengine/userspace.go:1104 +0x92
created by tailscale.com/ipn/ipnlocal.(*LocalBackend).requestEngineStatusAndWait
	/home/runner/work/tailscale/tailscale/ipn/ipnlocal/local.go:2454 +0xcf
exit status 2
FAIL	tailscale.com/ipn/ipnlocal	0.584s
```

(from https://github.com/tailscale/tailscale/pull/3019/checks?check_run_id=3831283855)

--- a/ipn/ipnlocal/state_test.go
+++ b/ipn/ipnlocal/state_test.go
@@ -17,6 +17,7 @@ import (
 	"tailscale.com/ipn"
 	"tailscale.com/ipn/store/mem"
 	"tailscale.com/tailcfg"
+	"tailscale.com/tstest"
 	"tailscale.com/types/empty"
 	"tailscale.com/types/key"
 	"tailscale.com/types/logger"
@@ -92,7 +93,6 @@ type mockControl struct {
 	opts       controlclient.Options
 	logfActual logger.Logf
 	statusFunc func(controlclient.Status)
-	preventLog atomic.Bool
 
 	mu          sync.Mutex
 	calls       []string
@@ -109,7 +109,7 @@ func newMockControl(tb testing.TB) *mockControl {
 }
 
 func (cc *mockControl) logf(format string, args ...any) {
-	if cc.preventLog.Load() || cc.logfActual == nil {
+	if cc.logfActual == nil {
 		return
 	}
 	cc.logfActual(format, args...)
@@ -282,7 +282,7 @@ func (cc *mockControl) UpdateEndpoints(endpoints []tailcfg.Endpoint) {
 func TestStateMachine(t *testing.T) {
 	c := qt.New(t)
 
-	logf := t.Logf
+	logf := tstest.WhileTestRunningLogger(t)
 	store := new(testStateStorage)
 	e, err := wgengine.NewFakeUserspaceEngine(logf, 0)
 	if err != nil {
@@ -297,7 +297,6 @@ func TestStateMachine(t *testing.T) {
 
 	cc := newMockControl(t)
 	cc.statusFunc = b.setClientStatus
-	t.Cleanup(func() { cc.preventLog.Store(true) }) // hacky way to pacify issue 3020
 
 	b.SetControlClientGetterForTesting(func(opts controlclient.Options) (controlclient.Client, error) {
 		cc.mu.Lock()
@@ -316,9 +315,6 @@ func TestStateMachine(t *testing.T) {
 	notifies.expect(0)
 
 	b.SetNotifyCallback(func(n ipn.Notify) {
-		if cc.preventLog.Load() {
-			return
-		}
 		if n.State != nil ||
 			n.Prefs.Valid() ||
 			n.BrowseToURL != nil ||
@@ -329,7 +325,6 @@ func TestStateMachine(t *testing.T) {
 			logf("\n(ignored) %v\n\n", n)
 		}
 	})
-	t.Cleanup(func() { b.SetNotifyCallback(nil) }) // hacky way to pacify issue 3020
 
 	// Check that it hasn't called us right away.
 	// The state machine should be idle until we call Start().
@@ -924,7 +919,7 @@ func TestStateMachine(t *testing.T) {
 }
 
 func TestEditPrefsHasNoKeys(t *testing.T) {
-	logf := t.Logf
+	logf := tstest.WhileTestRunningLogger(t)
 	store := new(testStateStorage)
 	e, err := wgengine.NewFakeUserspaceEngine(logf, 0)
 	if err != nil {
@@ -1006,7 +1001,7 @@ func (s *testStateStorage) sawWrite() bool {
 func TestWGEngineStatusRace(t *testing.T) {
 	t.Skip("test fails")
 	c := qt.New(t)
-	logf := t.Logf
+	logf := tstest.WhileTestRunningLogger(t)
 	eng, err := wgengine.NewFakeUserspaceEngine(logf, 0)
 	c.Assert(err, qt.IsNil)
 	t.Cleanup(eng.Close)

--- a/tstest/log.go
+++ b/tstest/log.go
@@ -145,3 +145,50 @@ func (ml *MemLogger) String() string {
 	defer ml.Unlock()
 	return ml.Buffer.String()
 }
+
+// WhileTestRunningLogger returns a logger.Logf that logs to t.Logf until the
+// test finishes, at which point it no longer logs anything.
+func WhileTestRunningLogger(t testing.TB) logger.Logf {
+	var (
+		mu   sync.RWMutex
+		done bool
+	)
+
+	logger := func(format string, args ...any) {
+		t.Helper()
+
+		mu.RLock()
+		defer mu.RUnlock()
+
+		if done {
+			return
+		}
+		t.Logf(format, args...)
+	}
+
+	// t.Cleanup is run before the test is marked as done, so by acquiring
+	// the mutex and then disabling logs, we know that all existing log
+	// functions have completed, and that no future calls to the logger
+	// will log something.
+	//
+	// We can't do this with an atomic bool, since it's possible to
+	// observe the following race:
+	//
+	//    test goroutine                goroutine 1
+	//    --------------                -----------
+	//                                  check atomic, testFinished = no
+	//    test finishes
+	//    run t.Cleanups
+	//    set testFinished = true
+	//                                  call t.Logf
+	//                                  panic
+	//
+	// Using a mutex ensures that all actions in goroutine 1 in the
+	// sequence above occur atomically, and thus should not panic.
+	t.Cleanup(func() {
+		mu.Lock()
+		defer mu.Unlock()
+		done = true
+	})
+	return logger
+}

